primitives/header-chain/src/lib.rs

```diff
@@ -110,3 +112,107 @@ impl<H, P> AncestryChecker<H, P> for () {
 		true
 	}
 }
+
+/// A simple ancestry checker which verifies ancestry by walking every header between `child` and
+/// `ancestor`.
+pub struct LinearAncestryChecker;
+
+impl<H: HeaderT> AncestryChecker<H, Vec<H>> for LinearAncestryChecker {
+	fn are_ancestors(ancestor: &H, child: &H, proof: &Vec<H>) -> bool {
+		// You can't be your own parent
+		if ancestor == child {
+			return false;
+		}
+
+		// Let's make sure that the given headers are actually in the proof
+		if let Some(first) = proof.first() {
+			if first != ancestor {
+				return false;
+			}
+		} else {
+			return false;
+		};
```

Comment:
It seems they are not necessarily cached, so might be worth to do `proof.iter().map(|h| h.hash()).collect()` first and only then do the comparison.

Reply:
> It seems they are not necessarily cached, so might be worth to do proof.iter().map(|h| h.hash()).collect() first and only then do the comparison.

I don't see the advantage of this. If we have a proof where the `first` and `last` headers are fine, but the middle isn't valid we would have ended up iterating over the entire collection needlessly instead of exiting early (current behaviour). Am I missing something?